### PR TITLE
Correct collision algorithm in RZ.

### PR DESCRIPTION
**Pull request by Yin-YinjianZhao**
In this PR, we will try to do a particle momentum rotation before calling the collision function, such that different azimuth positions of particles are also taken into account.

### CODE REVIEW

**Comment by Yin-YinjianZhao** [comment]
[input.txt](https://github.com/ECP-WarpX/WarpX/files/7489143/input.txt)
This input file is used to test the PR.
With this PR, collisions should not change particle velocities.
Without this PR, collisions will change particle velocities randomly.
Capture
As shown in the plot, `1`, `2`, ..., `5` are five sample particles casually picked among total 640 particles.
All particles are tested using this script.
[script.txt](https://github.com/ECP-WarpX/WarpX/files/7489157/script.txt)
It compares all particle velocities over time and show that all their velocities are not changed.

BTW, a small bug is fixed in `UpdateMomentumPerezElastic.H`.
I added a relative difference judgement at the beginning.
I.e., if the two particle's velocities are too close, no need to collide.
Previous judgement is too strict, and may lead to `nan`, such that the test mentioned before fails.

File: Examples/Tests/collision/inputs_rz [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+electron.charge = -q_e
+electron.mass = m_e
+electron.injection_style = "NUniformPerCell"
+electron.num_particles_per_cell_each_dim = 1 10 1
+electron.profile = constant
+electron.density = 1.0e21
+electron.momentum_distribution_type = parse_momentum_function
+electron.momentum_function_ux(x,y,z) = "1.0*x/sqrt(x*x+y*y)"
+electron.momentum_function_uy(x,y,z) = "1.0*y/sqrt(x*x+y*y)"
```
Comment (ax3l) [inline Examples/Tests/collision/inputs_rz]: There is a division by zero possibility here, you need to add something like to avoid singularities in the distribution:
```suggestion
electron.momentum_function_ux(x,y,z) = "if(x*x+y*y>0, 1.0*x/sqrt(x*x+y*y), 0)"
electron.momentum_function_uy(x,y,z) = "if(x*x+y*y>0, 1.0*y/sqrt(x*x+y*y), 0)"
```

File: Examples/Tests/collision/inputs_rz [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#################################
+############ COLLISION ##########
+#################################
+collisions.collision_names = collision
+collision.species = electron electron
+collision.CoulombLog = 15.9
+
+diagnostics.diags_names = diag
+diag.intervals = 10
+diag.diag_type = Full
```
Comment (ax3l) [inline Examples/Tests/collision/inputs_rz]: Undocumented, hard-coded expectation of the regression tests :see_no_evil: 
```suggestion
diagnostics.diags_names = diag1
diag1.intervals = 10
diag1.diag_type = Full
```

Ref.:
https://github.com/ECP-WarpX/WarpX/blob/13f6e984562a6d4981f7a1adc33e4dd3bc68e271/Regression/WarpX-tests.ini#L40-L42

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H [hunk 3]
```diff
@@ -107,18 +107,39 @@ void ElasticCollisionPerez (
...
+#endif
+
     // call UpdateMomentumPerezElastic()
     {
       int i1 = I1s; int i2 = I2s;
       for (int k = 0; k < amrex::max(NI1,NI2); ++k)
       {
+
+#if (defined WARPX_DIM_RZ)
+          T_R const theta = theta2[I2[i2]]-theta1[I1[i1]];
```
Comment (dpgrote) [inline Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:130]: Can you add some comments here explaining what this is doing and why it is needed?

Comment (Yin-YinjianZhao) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:130]: Sure.
In our 2D-RZ, when particles are having the same `vr`, without this PR, their velocities would change after collision with each other, because they may have different azimuth angles, but their velocities are not supposed to be changed.
To fix this, we find out the angle between the two particles, and rotate one particle to be in parallel with the other particle, such that their velocities won't change after collision if they have the same `vr`. Then, we need to rotate that particle back.

Comment (dpgrote) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:130]: Sorry if I missed discussion of this, but is there an understanding of why the particle's velocity changes if the rotation is not done? It seems that the particle velocities will be at arbitrary angles and the collision does not depend on the particle's theta, so it's not clear why it matters.

File: Examples/Tests/collision/analysis_collision_rz.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#! /usr/bin/env python
+
+# Copyright 2019-2020 Yinjian Zhao
```
Comment (ax3l) [inline Examples/Tests/collision/analysis_collision_rz.py]:
```suggestion
# Copyright 2019-2021 Yinjian Zhao
```

File: Examples/Tests/collision/analysis_collision_rz.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+#! /usr/bin/env python
```
Comment (ax3l) [inline Examples/Tests/collision/analysis_collision_rz.py]:
```suggestion
#!/usr/bin/env python
```

File: Examples/Tests/collision/analysis_collision_rz.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        px2 = ad['particle_momentum_x'].to_ndarray()
+        py2 = ad['particle_momentum_y'].to_ndarray()
+
+error = np.max( abs(px1-px2)+abs(py1-py2) )
+
+print('error = ', error)
+print('tolerance = ', tolerance)
+assert(error < tolerance)
+
+test_name = last_fn[:-9] # Could also be os.path.split(os.getcwd())[1]
```
Comment (ax3l) [inline Examples/Tests/collision/analysis_collision_rz.py]: Please use the syntax with the path split, that's more generic #2385 :)

File: Regression/WarpX-tests.ini [hunk 7]
```diff
@@ -1788,6 +1788,24 @@ analysisRoutine = Examples/Tests/collision/analysis_collision_2d.py
...
+buildDir = .
+inputFile = Examples/Tests/collision/inputs_rz
+runtime_params =
+dim = 2
+addToCompileString = USE_RZ=TRUE USE_PSATD=FALSE
+restartTest = 0
+useMPI = 1
+numprocs = 1
+useOMP = 1
+numthreads = 1
```
Comment (ax3l) [inline Regression/WarpX-tests.ini:1801]: Looking good :)

File: Regression/WarpX-tests.ini [hunk 8]
```diff
@@ -1788,6 +1788,24 @@ analysisRoutine = Examples/Tests/collision/analysis_collision_2d.py
...
+compileTest = 0
+doVis = 0
+compareParticles = 0
+analysisRoutine = Examples/Tests/collision/analysis_collision_rz.py
+aux1File = Regression/PostProcessingUtils/post_processing_utils.py
+tolerance = 1.e-30
```
Comment (ax3l) [inline Regression/WarpX-tests.ini]: That's stricter than double-precision, is that reasonable? :)

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H [hunk 9]
```diff
@@ -107,18 +107,44 @@ void ElasticCollisionPerez (
...
       int i1 = I1s; int i2 = I2s;
       for (int k = 0; k < amrex::max(NI1,NI2); ++k)
       {
+
+#if (defined WARPX_DIM_RZ)
+//When particles are having the same vr, their velocities would change after collision,
+//because they may have different azimuth angles, but their velocities are not supposed to be changed.
+//To fix this, we rotate one particle to be in parallel with the other particle,
+//such that their velocities won't change after collision if they have the same vr.
+//Then, we rotate that particle back.
```
Comment (ax3l) [inline Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H]: Functionally the same, but alignment is neat:
```suggestion
          /* When particles are having the same vr, their velocities would change after collision,
           * because they may have different azimuth angles, but their velocities are not supposed to be changed.
           * To fix this, we rotate one particle to be in parallel with the other particle,
           * such that their velocities won't change after collision if they have the same vr.
           * Then, we rotate that particle back. */
```

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H [hunk 10]
```diff
@@ -107,18 +107,44 @@ void ElasticCollisionPerez (
...
+          T_R const theta = theta2[I2[i2]]-theta1[I1[i1]];
+          T_R u1xbuf;
+          u1xbuf = u1x[I1[i1]];
```
Comment (ax3l) [inline Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H]: The compiler has an easier time optimizing this:
```suggestion
          T_R const u1xbuf = u1x[I1[i1]];
```

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H [hunk 11]
```diff
@@ -107,18 +107,44 @@ void ElasticCollisionPerez (
...
               u2x[ I2[i2] ], u2y[ I2[i2] ], u2z[ I2[i2] ],
               n1, n2, n12,
               q1, m1, w1[ I1[i1] ], q2, m2, w2[ I2[i2] ],
               dt, L, lmdD,
               engine);
+
+#if (defined WARPX_DIM_RZ)
+          u1xbuf = u1x[I1[i1]];
+          u1x[I1[i1]] = u1xbuf*std::cos(-theta) - u1y[I1[i1]]*std::sin(-theta);
+          u1y[I1[i1]] = u1xbuf*std::sin(-theta) + u1y[I1[i1]]*std::cos(-theta);
```
Comment (ax3l) [inline Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H]: The compiler has an easier time optimizing this:
```suggestion
          T_R const u1xbuf_new = u1x[I1[i1]];
          u1x[I1[i1]] = u1xbuf_new*std::cos(-theta) - u1y[I1[i1]]*std::sin(-theta);
          u1y[I1[i1]] = u1xbuf_new*std::sin(-theta) + u1y[I1[i1]]*std::cos(-theta);
```

**Review by ax3l — COMMENTED** [review]
Great work! Tiny suggestions inline.

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H (hunk 3 above)
Comment (Yin-YinjianZhao) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:130]: If particles have the same `vr` but different theta, their `vx` and `vy` are different, so the particles' velocities may point to different directions, and a collision between them would change their velocities. Although their velocities may only change slightly and oscillate around their initial values, as shown in the above plot, but we don't want them to change at all.
@RemiLehe Could you explain more if I missed something?

File: Regression/WarpX-tests.ini (hunk 8 above)
Comment (Yin-YinjianZhao) [reply Regression/WarpX-tests.ini]: When I tested it, I got even smaller than 1e-30, I got something like 1e-36.
I'm not sure, do you want to release it to a bigger tolerance?

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H (hunk 3 above)
Comment (dpgrote) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:130]: This sounds like a colocation problem - particles with similar r and z can collide even though they have different theta and are in fact far apart from each other. The fix here then is to temporarily map the particles into a common theta. Will this be problematic when using higher order RZ modes since the structure in theta is important? Could there be some theta dependence on the velocity distribution that this fix washes out? A harder question is that could there be a theta dependence in density that will affect collision rates?

Comment (RemiLehe) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:130]: @dpgrote Yes, I think that you are right for higher mode.
The fix implemented in this PR will be enough for one application that we have in mind in the short term, but you are right that it will cause problem with higher modes. I will add this topic on the agenda for the next developer meeting, and we could discuss it more.

Comment (Yin-YinjianZhao) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:130]: @dpgrote I'm not very familiar with higher order RZ modes, but I believe what you said is correct. This PR only works for pure 2D-RZ (no higher order modes). @RemiLehe What do you think? At least we should add some warning or assert in the code to make sure people don't use higher order RZ modes with this PR's fix.

Comment (dpgrote) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:130]: Ok. It would be good to add some comments here that this is assuming lowest mode only, i.e. axisymmetry. A warning would be good if the collisions are done with higher order modes turned on. Also, a statement should be added to the documentation for the collisions that it is assuming axisymmetry with RZ (in `usage/parameters.rst`).

File: Examples/Tests/collision/analysis_collision_rz.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+# This script tests the collision module in RZ.
+# Electrons are set with the same vr,
+# such that correct collisions will not lead to any velocity changes.
+# Thus the initial electron momentum will be compared with the final momentum.
+
+# Possible errors:
+# tolerance: 1.0e-30
+# Possible running time: ~ 1.0 s
+
+import sys
```
Comment (ax3l) [inline Examples/Tests/collision/analysis_collision_rz.py:20]:
```suggestion
import os
import sys
```

File: Regression/WarpX-tests.ini (hunk 8 above)
Comment (ax3l) [reply Regression/WarpX-tests.ini]: I wasn't sure if this is a relative or absolute tolerance.
Just checked and it is:
https://github.com/AMReX-Codes/regression_testing/blob/main/test_util.py#L144-L146

If we get a smaller tolerance in comparison out, then this is due to subnormal floating point numbers, which I would not necessarily rely on (e.g., on GPUs a common optimization is to flush subnormal floating point numbers to zero).

Long story short: I would go to set the tolerance to the regular double precision digits (1.e-15).

**Comment by RemiLehe** [comment]
As suggested by @NeilZaim, the generalization for higher-order modes could be implemented by binning particles in `theta`.

File: Docs/source/usage/parameters.rst [hunk 13]
```diff
@@ -1303,6 +1303,8 @@ Collision initialization
 
 WarpX provides a relativistic elastic Monte Carlo binary collision model,
 following the algorithm given by `Perez et al. (Phys. Plasmas 19, 083104, 2012) <https://doi.org/10.1063/1.4742167>`_.
+If dimension 2D-RZ is used, `warpx.n_rz_azimuthal_modes` must be set to 1 at the moment,
```
Comment (RemiLehe) [inline Docs/source/usage/parameters.rst]:
```suggestion
When the RZ mode is used, `warpx.n_rz_azimuthal_modes` must be set to 1 at the moment,
```

File: Docs/source/usage/parameters.rst [hunk 14]
```diff
@@ -1303,6 +1303,8 @@ Collision initialization
...
+i.e., axisymmetry is assumed.
```
Comment (RemiLehe) [inline Docs/source/usage/parameters.rst]:
```suggestion
since the current implementation of the collision module assumes axisymmetry.
```

File: Examples/Tests/collision/analysis_collision_rz.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+# Copyright 2019-2021 Yinjian Zhao
+#
+# This file is part of WarpX.
+#
+# License: BSD-3-Clause-LBNL
+
+# This script tests the collision module in RZ.
+# Electrons are set with the same vr,
+# such that correct collisions will not lead to any velocity changes.
```
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_rz.py]:
```suggestion
# and thus particles are traveling locally in the same direction. 
# Physically, the collision rate should therefore be very low. 
```

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H [hunk 16]
```diff
@@ -107,18 +107,48 @@ void ElasticCollisionPerez (
...
+           * Then, we rotate that particle back.
+           * Note that this only works for the lowest RZ mode, i.e., axisymmetry. */
+          AMREX_ALWAYS_ASSERT_WITH_MESSAGE(WarpX::n_rz_azimuthal_modes==1,
+            "RZ mode `warpx.n_rz_azimuthal_modes` must be 1 when using the binary collision module.");
+          T_R const theta = theta2[I2[i2]]-theta1[I1[i1]];
+          T_R const u1xbuf = u1x[I1[i1]];
+          u1x[I1[i1]] = u1xbuf*std::cos(theta) - u1y[I1[i1]]*std::sin(theta);
+          u1y[I1[i1]] = u1xbuf*std::sin(theta) + u1y[I1[i1]]*std::cos(theta);
+#endif
+
```
Comment (PhilMiller) [inline Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:135]: Why write to the backing arrays rather than just using local variables for these two values? This seems like pointless memory traffic.

Comment (PhilMiller) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:135]: Ok, so I see these are passed by reference, and updated inside `UpdateMomentumPerezElastic`. There's still an opportunity to do that write into a stack variable that won't be flushed to memory, and then only write it out once it's potentially been corrected by the following de-rotation.

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H [hunk 17]
```diff
@@ -107,18 +107,48 @@ void ElasticCollisionPerez (
...
+          /* When particles are having the same vr,
+           * their velocities would change after collision,
+           * because they may have different azimuth angles,
+           * but their velocities are not supposed to be changed.
+           * To fix this, we rotate one particle to be in parallel with the other particle,
+           * such that their velocities won't change after collision if they have the same vr.
+           * Then, we rotate that particle back.
+           * Note that this only works for the lowest RZ mode, i.e., axisymmetry. */
+          AMREX_ALWAYS_ASSERT_WITH_MESSAGE(WarpX::n_rz_azimuthal_modes==1,
+            "RZ mode `warpx.n_rz_azimuthal_modes` must be 1 when using the binary collision module.");
```
Comment (RemiLehe) [inline Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H]: This assert statement is inside a `for` loop that loops over cells (and potentially a GPU kernel), which maybe could affects performance (or readability of the error message in any case). 

Could you instead move this check to the code that initializes the collision module? For instance, here: https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/Collision/CollisionHandler.cpp#L32
(Note that the error should of course not be raised if there are not collisions at all, which is why I would suggest to add it inside the `for` loop over collision type, in the initialization code.)

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H [hunk 18]
```diff
@@ -107,18 +107,48 @@ void ElasticCollisionPerez (
...
+
+#if (defined WARPX_DIM_RZ)
+          /* When particles are having the same vr,
+           * their velocities would change after collision,
+           * because they may have different azimuth angles,
+           * but their velocities are not supposed to be changed.
+           * To fix this, we rotate one particle to be in parallel with the other particle,
+           * such that their velocities won't change after collision if they have the same vr.
+           * Then, we rotate that particle back.
+           * Note that this only works for the lowest RZ mode, i.e., axisymmetry. */
```
Comment (RemiLehe) [inline Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H]:
```suggestion
          /* In RZ geometry, macroparticles can collide with other macroparticles 
           * in the same *cylindrical* cell. For this reason, collisions between macroparticles
           * are actually not local in space. In this case, the underlying assumption is that 
           * particles within the same cylindrical cell represent a cylindrically-symmetry 
           * momentum distribution function. Therefore, here, we temporarily rotate the 
           * momentum of one of the macroparticles in agreement with this cylindrical symmetry.
           * (This is technically only valid if we use only the m=0 azimuthal mode in the simulation ; 
           * there is a corresponding assert statement at initialization.) */
```

**Review by RemiLehe — CHANGES_REQUESTED** [review]
Thanks for this PR. I added a request for the assert statement to be moved to another part of the code.

File: Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H (hunk 16 above)
Comment (Yin-YinjianZhao) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:135]: @PhilMiller Hi, thanks for you comments, could you provide more details how to improve the code, I'm not sure if I understand your point. Stack variables also work on GPUs?

Comment (PhilMiller) [reply Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H:135]: https://github.com/Yin-YinjianZhao/WarpX/pull/2
Here's a PR against your branch with my suggestion

**Review by RemiLehe — APPROVED** [review]
